src/cli/cli.cpp

```diff
@@ -5807,6 +5807,85 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
         SuccessOrExit(error = otMeshDiagDiscoverTopology(GetInstancePtr(), &config, HandleMeshDiagDiscoverDone, this));
         error = OT_ERROR_PENDING;
     }
+    /**
+     * @cli meshdiag childtable
+     * @code
+     * meshdiag childtable 0x6400
+     * rloc16:0x6402 ext-addr:8e6f4d323bbed1fe ver:4
+     *     timeout:120 age:36 supvn:129 q-msg:0
+     *     rx-on:yes type:ftd full-net:yes
+     *     rss - ave:-20 last:-20
+     *     err-rate - frame:11.51% msg:0.76%
+     *     csl - sync:no period:0 timeout:0
+     * rloc16:0x6403 ext-addr:ee24e64ecf8c079a ver:4
+     *     timeout:120 age:19 supvn:129 q-msg:0
+     *     rx-on:no type:mtd full-net:no
+     *     rss - ave:-20 last:-20
+     *     err-rate - frame:0.73% msg:0.00%
+     *     csl - sync:no period:0 timeout:0
+     * @endcode
+     * @par
+     * Start a query for child table of a router with a given RLOC16.
+     * Output lists all child entries. Information per child:
+     * * RLOC16
+     * * Extended MAC address
+     * * Thread Version
+     * * Timeout (in seconds)
+     * * Age (seconds since last heard)
+     * * Supervision interval (in seconds)
+     * * Number of queued messages (in case child is sleepy)
+     * * Device Mode
+     * * RSS (average and last)
+     * * Error rates: frame tx (at MAC layer), IPv6 message tx (above MAC)
+     * * CSL info
+     *   * If synchronized
+     *   * Period (in unit of 10-symbols-time)
+     *   * Timeout (in seconds)
+     * @cparam meshdiag childtable @ca{router-rloc16}
+     * @sa otMeshDiagQueryChildTable
+     */
+    else if (aArgs[0] == "childtable")
+    {
+        uint16_t routerRloc16;
+
+        SuccessOrExit(error = aArgs[1].ParseAsUint16(routerRloc16));
+        VerifyOrExit(aArgs[2].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
+
+        SuccessOrExit(error = otMeshDiagQueryChildTable(GetInstancePtr(), routerRloc16,
+                                                        HandleMeshDiagQueryChildTableResult, this));
+
+        error = OT_ERROR_PENDING;
+    }
+    /**
+     * @cli meshdiag childrenip6addrs
```

Comment:
Possible easier and shorter command name is `childip6` ?

Reply:
Good idea. Will rename it in next push. I will add more to this PR and will add "Answer TLV" for tracking multiple answers for the same query and will add a "Query ID TLV" so we can associate a query with its answer(s)`. It makes this PR bigger but I feel make sense to add and test all these related mechanisms together.

----

@EskoDijk want to also thank you for reviewing the PRs and specially paying attention to details and giving good suggestions. I do appreciate it. Thank you 😃 🙏

Reply:
@abtink Thanks! One thing still to consider is how would a Router with many children handle the multiple Answer messages? Frame buffer / queue space is limited and I think the Thread spec text prescribes the device to apply "backpressure" i.e. stop sending further Answers if the frame queue reaches a certain size or frame age (? not sure).

Ideally it would be best if the client would be in control using a request/response type protocol, as e.g. CoAP blockwise is also using - each next block is explicitly fetched by the client. That also reduces DoS opportunities e.g. sending a small request frame could cause an avalanche of data on the mesh.

Further thoughts: Also in CoAP blockwise, instead of a "Query ID TLV" the Token value is used as the identifier of the transaction. Not sure if we really need to reinvent all such mechanisms here?   The main problem of "Blockwise" is that it assumes one data blob is fetched and that data can be arbitrarily divided into blocks of certain sizes (32/64/128 etc).  That's not compatible with our model here.

Reply:
> One thing still to consider is how would a Router with many children handle the multiple Answer messages? Frame buffer / queue space is limited and I think the Thread spec text prescribes the device to apply "backpressure" i.e. stop sending further Answers if the frame queue reaches a certain size or frame age (? not sure).

Two ideas on this:
- First, we can rely on `Coap` module itself since it keeps the messages to retry if they fail, so if msg is dropped due to back-pressure logic, it will be retried by `Coap` module.
- Second one which I want to try out is to handle this in `NetworkDiag::Server` module itself: It sends the first answer and then waits for CoAP ack for it before sending the next answer (stop sending answer if no ack for a previous answer). I like this better for sure (I think it requires some changes in `Coap` module to add a callback for "AckRx" to use with confirmable CoAP msgs).

Reply:
> Also in CoAP blockwise, instead of a "Query ID TLV" the Token value is used as the identifier of the transaction. Not sure if we really need to reinvent all such mechanisms here? The main problem of "Blockwise" is that it assumes one data blob is fetched and that data can be arbitrarily divided into blocks of certain sizes (32/64/128 etc). That's not compatible with our model here.

I think adding "Query ID TLV" can help with multicast queries as well (same node sending multiple queries). 
We cannot use CoAP block-wise for query/answer model. We have an implementation of block-wise transfer in Coap, but it is not really enabled on devices and/or tested much (it is intended for experimental use by application developers). It adds some good chunk of code overhead (~5-6 KB of code when enabled)

Reply:
Agree, something specific in this case would be better than forcing CoAP Blockwise (which doesn't fit our use case very well).

On the 2nd idea of waiting for the CoAP ACK: this Ack, in response to a TMF Answer, is in fact (or should  be) a "piggybacked response" which contains the Ack with the 2.04 response. The latter (response code) is for sure something that the CoAP module ought to expose to the next higher layer. Because it is really an application-level concept.  So at CoAP module level it would make perfect sense to expose this response code and use a callback or so to get its value back to the layer using CoAP.  

Exposing the ACK itself in this way would make less sense, I think, this is really a CoAP-internal layer used for the reliability.  At least normally the application code doesn't need to be informed of all the details of retransmissions or notifying the application that the request was received - which the Ack is about.  But because of the piggybacked communication pattern the Ack and the REsponse tend to get named interchangably, but in principle these are different concepts.

One added benefit of the diagnostic layer getting back the CoAP response code from the entity requesting the data is that it can be used for control of the remainder of Answers:  for example if the requesting entity uses 2.xx success response then the Router can continue sending more Answers. If the entity uses an error code (4.xx / 5.xx) then this may signal the Router to stop sending further Answers.

Reply:
Section 10.4.4 of the spec btw lists the current CoAP response codes usable for TMF.  In principle these can equally be used for particular error cases instead of the "dummy 2.04" response we currently define for the TMF Answer.

Reply:
@abtink In CoAP, you would always know which CoAP request a CoAP response belongs to: the Token value. That means if a node is sending multiple TMF Answer messages (=CoAP requests) at the same time and to the same destination, it can still distinguish the different CoAP responses (e.g. all 2.04) coming in by their Token value.

But maybe the problem you indicate is one layer above CoAP? E.g. if the application layer doesn't see the CoAP Token because the CoAP layer adds it, then the application cannot match requests and responses using the Token value. Then the API for the "send CoAP request" call could immediately return a handle (ID, or pointer) to identify this request and the same handle is used for responses coming in.  Or the other solution of a context which achieves the same I think.

Reply:
> But maybe the problem you indicate is one layer above CoAP?
>  Or the other solution of a context which achieves the same I think.

Yes Exactly. :). I was searching for a way to realize this without needing to track anything in `NetDiag` module and/or need to change the existing CoAP methods and callbacks.